specifyweb/specify/tree_views.py

```diff
@@ -343,4 +351,392 @@ def perm_target(tree):
+
+        # If there are no ranks lower than the target rank, then add the new rank to the top of the hierarchy
+        elif is_new_rank_last:
+            max_rank_id = rank_ids[-1]
+            new_rank_id = max_rank_id + rank_increment # TODO: checkout
+
+        # If the new rank is being placed somewhere in the middle of the heirarchy
+        else:
+            new_rank_id = int((next_rank_id - parent_rank_id) / 2) + parent_rank_id
+            if next_rank_id - parent_rank_id < 1:
+                return HttpResponseBadRequest(f"Can't add rank id between {new_rank_id} and {parent_rank_id}")
+
+    # Create and save the new TreeDefItem record
+    new_fields_dict['rankid'] = new_rank_id
+    new_rank = tree_def_item_model.objects.create(**new_fields_dict)
+    new_rank.save()
+
+    # Set the parent rank, that previously pointed to the target, to the new rank
+    child_ranks = tree_def_item_model.objects.filter(treedef=tree_def, parent=parent_rank).exclude(rankid=new_rank_id)
+    if child_ranks.exists():
+        # Iterate through the child ranks, but there should only ever be 0 or 1 child ranks to update
+        for child_rank in child_ranks:
+            child_rank.parent = new_rank 
+            child_rank.save()
+
+    # Regenerate full names
+    tree_extras.set_fullnames(tree_def, null_only=False, node_number_range=None)
+
+    logger.info(f"Added new tree rank: {new_rank_name} with ID: {new_rank_id}")
+    return HttpResponse("Success")
+
+@openapi(schema={
+    'post': {
+        "requestBody": {
+            "required": True,
+            "description": "Delete a rank from a tree.",
+            "content": {
+                "application/json": {
+                    "schema": {
+                        "type": "object",
+                        "description": "The request body.",
+                        "properties": {
+                            "rankName": {
+                                "type": "string",
+                                "description": "The name of the rank to delete."
+                            },
+                            "treeName": {
+                                "type": "string",
+                                "description": "The name of tree."
+                            },
+                            "treeID": {
+                                "type": "integer",
+                                "description": "The ID of the tree."
+                            }
+                        },
+                        'required': ['rankName'],
+                        'additionalProperties': False
+                    }
+                }
+            }
+        },
+        "responses": {
+            "200": {"description": "Success"},
+            "400": {"description": "Cannot delete tree rank"},
+            "500": {"description": "Server Error"}
+        }
+    },
+})
+@login_maybe_required
+@require_POST
+@transaction.atomic
+def delete_tree_rank(request, tree) -> HttpResponse:
```

Comment:
does this endpoint have to exist?
could you just use the `@orm_signal_handler('pre_delete', 'TaxonTreeDefItem')` (and post_delete, for each tree) to automatically clean up before and after tree rank deletion?